podpac/datalib/soilscape.py

```diff
@@ -0,0 +1,669 @@
+        except:
+            _logger.exception("Could not get coordinates for '%s' node '%s'" % (NODE2SITE[node], node))
+            continue
+        lats.append(lat)
+        lons.append(lon)
+
+    return podpac.Coordinates([[lats, lons], time, depth], dims=["lat_lon", "time", "alt"], crs=CRS)
+
+
+class SoilSCAPEFile(podpac.data.Dataset):
+    """ SoilSCAPE datasource from file. """
+
+    data_key = ["soil_moisture", "moisture_flag"]
+    alt_key = "depth"
+
+    @property
+    def lat(self):
+        return self.dataset.lat.item()
+
+    @property
+    def lon(self):
+        return self.dataset.lon.item()
+
+    @property
+    def physicalid(self):
+        # note: this should be the same as the node number
+        return self.dataset.physicalid.item()
+
+    def get_coordinates(self):
+        coordinates = super(SoilSCAPEFile, self).get_coordinates()
+        coordinates.set_trait("crs", CRS)
+        return coordinates
+
+
+class SoilSCAPENode(SoilSCAPEFile):
+    """ SoilSCAPE 20min soil moisture for a particular node.
+
+        Data is loaded from the THREDDS https fileserver.
+
+        Attributes
+        ----------
+        site : str
+            SoilSCAPE site, e.g. 'Canton_OK'.
+        node : int
+            SoilSCAPE node id.
+    """
+
+    site = tl.Enum(list(NODES)).tag(attr=True)
+    node = tl.Int().tag(attr=True)
+    cache_dataset = tl.Bool(True)
+
+    _repr_keys = ["site", "node"]
+
+    @tl.validate("node")
+    def _validate_node(self, d):
+        if d["value"] not in NODES[self.site]:
+            raise ValueError("Site '%s' does not have a node n%d" % (self.site, d["value"]))
+
+        return d["value"]
+
+    @property
+    def source(self):
+        return "{base_url}/{filename}.nc".format(base_url=SOILSCAPE_FILESERVER_BASE, filename=self.filename)
+
+    @property
+    def filename(self):
+        return "soil_moist_20min_{site}_n{node}".format(site=self.site, node=self.node)
+
+
+class SoilSCAPE20min(podpac.core.compositor.compositor.BaseCompositor):
+    """ SoilSCAPE 20min soil moisture data for an entire site.
+
+        Data is loaded from the THREDDS https fileserver.
+
+        Attributes
+        ----------
+        site : str
+            SoilSCAPE site, e.g. 'Canton_OK'.
+        exclude : list
+            data points with these quality flags will be excluded. Default excludes [1, 2, 3, 4].
+            Flags::
+             * 0 - (G) Good (Standard for all data)
+             * 1 - (D) Dubious (Automatically flagged, spikes etc.,)
+             * 2 - (I) Interpolated / Estimated
+             * 3 - (B) Bad (Manually flagged)
+             * 4 - (M) Missing
+             * 5 - (C) Exceeds field size (Negative SM values, fixed at 0.1 percent)
+    """
+
+    site = tl.Enum(list(NODES), allow_none=True, default_value=None).tag(attr=True)
+    exclude = tl.List([1, 2, 3, 4]).tag(attr=True)
+    dataset_expires = tl.Any()
+
+    @tl.default("dataset_expires")
```

Comment:
Here is the cache expiration in use. I'm not sure if 1 day is the right expiration time. The THREDDS data doesn't seem to be updated that very frequently, or recently at all. But this is a reliable starting point.

We could also remove the default and only specify an expiration in user-code that needs it (i.e. that actually specifies a cache_ctrl, turns on cache_dataset, and wants to get the latest possible data frequently).

Reply:
@mpu-creare Okay, actually I removed the default, but it is passed through to the dataset nodes, so a user can specify when necessary. I think that is cleaner.